runtime: Add "none" as a shared_fs option

fidencio opened the pull request:
Currently, even when using devmapper, if the VMM supports virtio-fs /
virtio-9p, that's used to share a few files between the host and the
guest.

This needed, as we need to share with the guest contents like secrets,
certificates, and configurations, via Kubernetes objects like configMaps
or secrets, and those are rotated and must be updated into the guest
whenever the rotation happens.

However, there are still use-cases users can live with just copying
those files into the guest at the pod creation time, and for those
there's absolutely no need to have a shared filesystem process running
with no extra obvious benefit, consuming memory and even increasing the
attack surface used by Kata Containers.

For the case mentioned above, we should allow users, making it very
clear which limitations it'll bring, to run Kata Containers with
devmapper without actually having to use a shared file system, which is
already the approach taken when using Firecracker as the VMM.

Fixes: #7207

fidencio commented:
I'll split this patch, but I'd like to get some initial feedback on it.

wedsonaf commented:
I like this. We'll need something like that for CC when running with CLH.

I was using the `ConfidentialGuest` config to disable fs sharing in CLH (https://github.com/microsoft/kata-containers/commit/3631bd3c11e61b5b1a3db9b61d44e99f0568eb0e) but I like this explicit option better, especially because we may eventually have other FS in clh-cc in the future.

fidencio commented on src/runtime/pkg/device/config/config.go:
```diff
@@ -81,6 +81,9 @@ const (
 	VirtioFSNydus = "virtio-fs-nydus"
+
+	// VirtioFSNone means files will be copied into the guest system
+	VirtioFSNone = "none"
```
Totally, and we should also make it very clear in the configuration file.

egernst replied:
Specifically state that this is static at pod creation time, and that any updates to the mount will not be reflected during life cycle of the pod. As an example, this will result in rotation of secrets, certificates or configurations via kubernetes objects like configMaps or secrets.

wedsonaf commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -1009,15 +1009,16 @@ func newClhHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
-	}
-
-	if h.VirtioFSDaemon == "" {
+	if (sharedFS == config.VirtioFS || sharedFS != config.VirtioFSNydus) && h.VirtioFSDaemon == "" {
```
Did you mean `sharedFs == config.VirtioFSNydus` here?

fidencio replied:
I did. :-)

wedsonaf commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -1009,15 +1009,16 @@ func newClhHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
 	}
 
+
+	if sharedFS == config.Virtio9P {
```
The change you're making here is analogous to going from an allow-list to a deny-list.

I think the consensus is that an allow-list is better though.

In this particular case, suppose we add a new Fs to kata that clh doesn't support; with a deny-list, we would wrongly accept it. With an allow-list, we would reject it. I think the latter is preferable because if it is meant to be allowed, someone will notice it and just add it here. Allowing something that shouldn't be allowed is harder to catch.

fidencio replied:
Sure, I will change the check back to how it was, and also re-write the check above.

bergwolf commented on src/runtime/virtcontainers/qemu_arch_base.go:
```diff
@@ -168,12 +168,13 @@ type qemuArchBase struct {
+	disableNvdimm   bool
+	dax             bool
+	legacySerial    bool
+	disableSharedFS bool
```
This is unused?

amshinde commented on src/runtime/pkg/device/config/config.go:
```diff
@@ -81,6 +81,9 @@ const (
 	VirtioFSNydus = "virtio-fs-nydus"
+
+	// VirtioFSNone means files will be copied into the guest system
+	VirtioFSNone = "none"
```
I am fine with naming this as none, but was wondering if making as this "copy-files" instead of  "none" would be benefecial in terms of adding clarity.

fidencio commented:
Folks, this one is ready for review and I think I've addressed all your comments.

Please, take a second look at the PR whenever it's possible.

I've tested this locally with both Cloud Hypervisor and QEMU, which are the affected VMMs.

fidencio commented:
/test

fidencio commented:
/test

bergwolf reviewed (approved):
lgtm!

egernst commented on src/runtime/pkg/device/config/config.go:
```diff
@@ -81,6 +81,17 @@ const (
+	// secrets, certs, or configurations via kubernetes objects like
+	// configMaps or secrets, as those will be copied into the guest at
+	// *pod* *creation* *time*.
+	VirtioFSNone = "none"
```
Why string?

egernst commented on src/runtime/pkg/katautils/config.go:
```diff
@@ -1009,11 +1009,12 @@ func newClhHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
 
-	if sharedFS != config.VirtioFS && sharedFS != config.VirtioFSNydus {
-		return vc.HypervisorConfig{}, errors.New("clh only support virtio-fs or virtio-fs-nydus")
+	if sharedFS != config.VirtioFS && sharedFS != config.VirtioFSNydus && sharedFS != config.VirtioFSNone {
```
This is a bit awkward, in that we also are doing just about this same check up above in the map of supported shared-filesystem options in sharedFS.  I wonder if there's a way we can dedup/clean this up, so we don't have to make changes in multiple spots again next time.

egernst replied:
Ah, sorry - I see that's a 'general' check, whereas this is cloud hypervisor specific check.  meh, its a bit awkward/broken abstraction that we are checking clh specific bits in a generic katautils pkg, but I guess that isn't necessarily your fault.

egernst commented:
This PR reminded me of a couple of broken abstractions and how large VC itself is, but neither item is your fault here.  The changes lgtm @fidencio. Are there integration tests that exercise this available already? I'd prefer we have these queued up and exercised -- I trust you tested, but I don't trust that we won't immediately break it if we don't have tests.

fidencio commented:
> Are there integration tests that exercise this available already?

@egernst, there are tests for the capaibilites on the hypervisor side, and I expanded those to ensure that whenever "NoSharedFS" is set, "IsFsSharingSupported" returns false.

I've expanded this on the QEMU side, and added it to the Cloud Hypervisor side.

Please, let me know if this is enough to cover your concerns.

egernst reviewed (approved):
Thanks - I like the rename - reads more natural now.  lgtm.

fidencio commented:
/test
